File: integrations/langgraph/python/examples/agents/a2ui_dynamic_schema/agent.py
"""
Dynamic A2UI tool: LLM-generated UI from conversation context.

A secondary LLM generates v0.9 A2UI components via a structured tool call.
The generate_a2ui tool wraps the output as a2ui_operations, which the
middleware detects in the TOOL_CALL_RESULT and renders automatically.
"""

import json
import os
from typing import Any, List

from langchain.tools import tool, ToolRuntime
from langchain_core.messages import SystemMessage
from langchain_core.tools import tool as lc_tool
from langchain_core.runnables import RunnableConfig
from langchain_openai import ChatOpenAI
from langgraph.graph import StateGraph, END, MessagesState
from langgraph.prebuilt import ToolNode

from copilotkit import a2ui


@lc_tool
def render_a2ui(
    surfaceId: str,
    catalogId: str,
    components: list[dict],
    data: dict | None = None,
) -> str:
    """Render a dynamic A2UI v0.9 surface.

    Args:
        surfaceId: Unique surface identifier.
        catalogId: The catalog ID (use "https://a2ui.org/demos/dojo/custom_catalog.json").
        components: A2UI v0.9 component array (flat format). The root
            component must have id "root".
        data: Optional initial data model for the surface (e.g. form values,
            list items for data-bound components).
    """
    return "rendered"


def _build_context_prompt(state: dict) -> str:
    """Build the A2UI generation prompt from client-provided context entries.

    The frontend sends generation guidelines, design guidelines, and the
    component schema as separate context entries. The LangGraph integration
    also extracts the schema into state["ag-ui"]["a2ui_schema"].
    """
    ag_ui = state.get("ag-ui", {})
    parts: list[str] = []

    # Include all context entries (generation guidelines, design guidelines, etc.)
    # Entries may be Pydantic Context objects or plain dicts.
    for entry in ag_ui.get("context", []):
        desc = entry.description
        value = entry.value
        if desc:
            parts.append(f"## {desc}\n{value}\n")
        else:
            parts.append(f"{value}\n")

    # Include A2UI component schema (separated out by the LangGraph integration)
    a2ui_schema = ag_ui.get("a2ui_schema")
    if a2ui_schema:
        parts.append(f"## Available Components\n{a2ui_schema}\n")

    return "\n".join(parts)


CUSTOM_CATALOG_ID = "https://a2ui.org/demos/dojo/dynamic_catalog.json"

# Local composition guide — tells the secondary LLM how to use our
# pre-made domain components (HotelCard, ProductCard, TeamMemberCard).
COMPOSITION_GUIDE = """
## Available Pre-made Components

You have 4 components. Use Row as the root with structural children to repeat a card per item.

### Row
Layout container. Use structural children to repeat a card template:
  {"id":"root","component":"Row","children":{"componentId":"card","path":"/items"}}

### HotelCard
Props: name, location, rating (number 0-5), pricePerNight, amenities (optional), action
Example:
  {"id":"card","component":"HotelCard","name":{"path":"name"},"location":{"path":"location"},
   "rating":{"path":"rating"},"pricePerNight":{"path":"pricePerNight"},
   "action":{"event":{"name":"book","context":{"name":{"path":"name"}}}}}

### ProductCard
Props: name, price, rating (number 0-5), description (optional), badge (optional), action
Example:
  {"id":"card","component":"ProductCard","name":{"path":"name"},"price":{"path":"price"},
   "rating":{"path":"rating"},"description":{"path":"description"},
   "action":{"event":{"name":"select","context":{"name":{"path":"name"}}}}}

### TeamMemberCard
Props: name, role, department (optional), email (optional), avatarUrl (optional), action
Example:
  {"id":"card","component":"TeamMemberCard","name":{"path":"name"},"role":{"path":"role"},
   "department":{"path":"department"},"email":{"path":"email"},
   "action":{"event":{"name":"contact","context":{"name":{"path":"name"}}}}}

## RULES
- Root is ALWAYS a Row with structural children: {"componentId":"<card-id>","path":"/items"}
- Inside templates, use RELATIVE paths (no leading slash): {"path":"name"} not {"path":"/name"}
- Always provide data in the "data" argument as {"items":[...]}
- Pick the card type that best matches the user's request
- Generate 3-4 realistic items with diverse data
"""


@tool()
def generate_a2ui(runtime: ToolRuntime[Any]) -> str:
    """Generate dynamic A2UI components based on the conversation.

    A secondary LLM designs the UI schema and data. The result is
    returned as an a2ui_operations container for the middleware to detect.
    """
    # The last message is this tool call (generate_a2ui) so we remove it,
    # as it is not yet balanced with a tool call response.
    messages = runtime.state["messages"][:-1]

    # Build prompt from client-provided context + local composition guide
    prompt = _build_context_prompt(runtime.state) + "\n" + COMPOSITION_GUIDE

    model = ChatOpenAI(model="gpt-4.1")
    model_with_tool = model.bind_tools(
        [render_a2ui],
        tool_choice="render_a2ui",
    )

    response = model_with_tool.invoke(
        [SystemMessage(content=prompt), *messages],
    )

    # Extract the render_a2ui tool call arguments
    if not response.tool_calls:
        return json.dumps({"error": "LLM did not call render_a2ui"})

    tool_call = response.tool_calls[0]
    args = tool_call["args"]

    surface_id = args.get("surfaceId", "dynamic-surface")
    catalog_id = args.get("catalogId", CUSTOM_CATALOG_ID)
    components = args.get("components", [])
    data = args.get("data", {})

    # Wrap as v0.9 a2ui_operations so the middleware detects it
    ops = [
        a2ui.create_surface(surface_id, catalog_id=catalog_id),
        a2ui.update_components(surface_id, components),
    ]
    if data:
        ops.append(a2ui.update_data_model(surface_id, data))

    result = a2ui.render(operations=ops)
    return result


TOOLS = [generate_a2ui]


class AgentState(MessagesState):
    tools: List[Any]
    copilotkit: dict  # CopilotKit context (actions, etc.)

# LangGraph requires state keys declared in the schema.
# "ag-ui" uses a hyphen which isn't valid as a Python identifier,
# so we patch it into the annotations directly.
AgentState.__annotations__["ag-ui"] = dict


SYSTEM_PROMPT = """You are a helpful assistant that creates rich visual UI on the fly.

When the user asks for visual content (product comparisons, dashboards, lists, cards, etc.),
use the generate_a2ui tool to create a dynamic A2UI surface.
IMPORTANT: After calling the tool, do NOT repeat the data in your text response. The tool renders UI automatically. Just confirm what was rendered."""


async def chat_node(state: AgentState, config: RunnableConfig):
    model = ChatOpenAI(model="gpt-4o")
    model = model.bind_tools(TOOLS, parallel_tool_calls=False)

    response = await model.ainvoke([
        SystemMessage(content=SYSTEM_PROMPT),
        *state["messages"],
    ], config)

    return {"messages": [response]}


def route_after_chat(state: AgentState):
    last_message = state["messages"][-1]
    if hasattr(last_message, "tool_calls") and last_message.tool_calls:
        return "tool_node"
    return END


workflow = StateGraph(AgentState)
workflow.add_node("chat_node", chat_node)
workflow.add_node("tool_node", ToolNode(tools=TOOLS))
workflow.set_entry_point("chat_node")
workflow.add_conditional_edges("chat_node", route_after_chat)
workflow.add_edge("tool_node", "chat_node")

is_fast_api = os.environ.get("LANGGRAPH_FAST_API", "false").lower() == "true"

if is_fast_api:
    from langgraph.checkpoint.memory import MemorySaver
    memory = MemorySaver()
    graph = workflow.compile(checkpointer=memory)
else:
    graph = workflow.compile()

File: integrations/langgraph/typescript/examples/src/agents/a2ui_recovery/agent.ts
/**
 * A2UI recovery agent (OSS-162) — DRAFT showcase, verify before wiring.
 *
 * Mirrors `a2ui_dynamic_schema` but enables the error-recovery loop: passes a
 * `catalog` (so the sub-agent's output is validated against component schemas)
 * and a `recovery` config to `getA2UITools`. When the sub-agent emits an invalid
 * A2UI surface (e.g. a HotelCard missing its required `rating`), the validation
 * errors are fed back and it regenerates, up to `maxAttempts`; the middleware
 * suppresses faulty attempts (no wipe) and surfaces an `a2ui_recovery` status.
 *
 * In the dojo demo the sub-agent's render_a2ui output is driven deterministically
 * by aimock (invalid → valid), so recovery is reproducible without a real LLM.
 *
 * ⚠️ For the gate to fire on SEMANTIC errors, the SAME `catalog` must also reach
 * `@ag-ui/a2ui-middleware` (its `schema` option). See INTEGRATION-CHECKLIST.
 */

import { createAgent } from "langchain";
import { copilotkitMiddleware } from "@copilotkit/sdk-js/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import { getA2UITools } from "@ag-ui/langgraph";
import type { A2UIValidationCatalog } from "@ag-ui/a2ui-toolkit";

const CUSTOM_CATALOG_ID = "https://a2ui.org/demos/dojo/dynamic_catalog.json";

// Catalog (component name → required props) used to validate the sub-agent's
// output. Must match the dojo dynamic catalog (apps/dojo/src/a2ui-catalog) and
// the `schema` handed to A2UIMiddleware.
const RECOVERY_CATALOG: A2UIValidationCatalog = {
  components: {
    Row: { required: ["children"] },
    HotelCard: { required: ["name", "location", "rating", "pricePerNight", "action"] },
    ProductCard: { required: ["name", "price", "rating", "action"] },
    TeamMemberCard: { required: ["name", "role", "action"] },
  },
};

const COMPOSITION_GUIDE = `
## Available Pre-made Components

You have 4 components. Use Row as the root with structural children to repeat a card per item.

### Row
Layout container. Use structural children to repeat a card template:
  {"id":"root","component":"Row","children":{"componentId":"card","path":"/items"}}

### HotelCard
Props (ALL required unless noted): name, location, rating (number 0-5), pricePerNight, action; amenities (optional)

### ProductCard
Props: name, price, rating (number 0-5), action; description (optional), badge (optional)

### TeamMemberCard
Props: name, role, action; department (optional), email (optional), avatarUrl (optional)

## RULES
- Root is ALWAYS a Row with structural children: {"componentId":"<card-id>","path":"/items"}
- Inside templates, use RELATIVE paths (no leading slash): {"path":"name"} not {"path":"/name"}
- Always provide data in the "data" argument as {"items":[...]}
- Include EVERY required prop on each card.
- Generate 3-4 realistic items with diverse data.
`;

const a2uiTool = getA2UITools(new ChatOpenAI({ model: "gpt-4o" }), {
  defaultCatalogId: CUSTOM_CATALOG_ID,
  compositionGuide: COMPOSITION_GUIDE,
  // OSS-162: enable catalog-aware recovery.
  catalog: RECOVERY_CATALOG,
  recovery: { maxAttempts: 3 },
  onA2UIAttempt: (rec) => {
    // Dev observability: each attempt (incl. rejected ones) is logged.
    // eslint-disable-next-line no-console
    console.log(`[a2ui recovery] attempt ${rec.attempt}: ${rec.ok ? "valid" : "invalid"}`, rec.errors);
  },
});

export const a2uiRecoveryGraph = createAgent({
  model: "openai:gpt-4o",
  // Cast: tool typed against @ag-ui/langgraph's own @langchain/core peer.
  tools: [a2uiTool as any],
  middleware: [copilotkitMiddleware],
  systemPrompt: `You are a helpful assistant that creates rich visual UI on the fly.

When the user asks for visual content (hotel/product comparisons, team rosters, lists, cards, etc.),
use the generate_a2ui tool to create a dynamic A2UI surface.
IMPORTANT: After calling the tool, do NOT repeat the data in your text response. The tool renders UI automatically. Just confirm what was rendered.`,
});
